storey/sources.py

```diff
@@ -1012,3 +1012,99 @@ def _init(self):
+    """Use SQL table as input source for a flow.
+
+
+    :parameter key_field: the primary key of the table.
+    :parameter time_field: column to be used as time for events.
+    :parameter id_field: column to be used as ID for events.
+    :parameter db_name: url string connection to sql database.
+    :parameter table_name: the name of the table to access,
+                                from the current database
+
+    """
+
+    def __init__(
+        self,
+        db_path: str = None,
+        table_name: str = None,
+        key_field: Optional[Union[str, List[str]]] = None,
+        time_field: Optional[str] = None,
+        id_field: Optional[str] = None,
+        **kwargs,
+    ):
+
+        if key_field is not None:
+            kwargs["key_field"] = key_field
+        if time_field is not None:
+            kwargs["time_field"] = time_field
+        if id_field is not None:
+            kwargs["id_field"] = id_field
+        _IterableSource.__init__(self, **kwargs)
+        WithUUID.__init__(self)
+
+        if not all([db_path, table_name]):
+            raise ValueError("cannot specify without db_path and table_name args")
+
+        self.table_name = table_name
+        self.db_path = db_path
+
+        self._key_field = key_field
+        if isinstance(time_field, str) and "." in time_field:
+            self._time_field = time_field.split(".")
+        else:
+            self._time_field = time_field
+        if isinstance(id_field, str) and "." in id_field:
+            self._id_field = id_field.split(".")
+        else:
+            self._id_field = id_field
+
+    async def _run_loop(self):
+        import sqlalchemy as db
+
+        engine = db.create_engine(self.db_path)
+        metadata = db.MetaData()
+        with engine.connect() as conn:
+            table = db.Table(self.table_name, metadata, autoload=True, autoload_with=engine)
+            results = conn.execute(db.select([table])).fetchall()
+
+        for namedtuple in results:
+            create_event = True
+            body = namedtuple._asdict()
+            body.pop("index")
+            key = None
+            if self._key_field:
+                if isinstance(self._key_field, list):
+                    key = []
+                    for key_field in self._key_field:
+                        if key_field not in body or pandas.isna(body[key_field]):
+                            create_event = False
+                            break
+                        key.append(body[key_field])
+                else:
+                    key = body[self._key_field]
+                    if key is None:
+                        create_event = False
+            if create_event:
+                time = None
+                if self._time_field:
+                    time = self.get_val_from_multi_dictionary(body, self._time_field)
+                if self._id_field:
+                    _id = self.get_val_from_multi_dictionary(body, self._id_field)
```

Comment:
BUG. These two ifs are not covered by any tests. It they were, the test would fail because this code, instead of accessing field "my_id", will try to access fields "m", "y", "_", "i", and "d".

Reply:
Afaict, this code was copied from `DataframeSource` and then the bug was added. Please try to avoid duplicating this code.